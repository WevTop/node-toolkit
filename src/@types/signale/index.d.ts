import 'signale'

declare module 'signale' {
  interface SignaleBase<TTypes extends string = DefaultMethods> {
    public disable (): void
  }
}
