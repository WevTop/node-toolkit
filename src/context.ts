export default class Context {
  public payload: any
  public sha: string
  public ref: string

  constructor () {
    this.payload = require(String(process.env.GITHUB_EVENT_PATH))
    this.sha = String(process.env.GITHUB_SHA)
    this.ref = String(process.env.GITHUB_REF)
  }

  /**
   * Return the `owner` and `repo` params for making API requests against a
   * repository.
   *
   * ```js
   * const params = context.repo({path: '.github/config.yml'})
   * // Returns: {owner: 'username', repo: 'reponame', path: '.github/config.yml'}
   * ```
   *
   * @param object - Params to be merged with the repo params.
   *
   */
  repo (object: object) {
    const repo = this.payload.repository

    if (!repo) {
      throw new Error('toolkit.context.repo() is not supported for this webhook event.')
    }

    return {
      owner: repo.owner.login || repo.owner.name,
      repo: repo.name,
      ...object
    }
  }

  /**
   * Return the `owner`, `repo`, and `number` params for making API requests
   * against an issue or pull request. The object passed in will be merged with
   * the repo params.
   *
   * ```js
   * const params = context.issue({body: 'Hello World!'})
   * // Returns: {owner: 'username', repo: 'reponame', number: 123, body: 'Hello World!'}
   * ```
   *
   * @param object - Params to be merged with the issue params.
   */
  issue (object: object) {
    const payload = this.payload
    return this.repo({
      number: (payload.issue || payload.pull_request || payload).number,
      ...object
    })
  }
}
