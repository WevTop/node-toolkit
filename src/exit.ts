// tslint:disable:no-console

/**
 * The code to exit an action with a "success" state
 */
export const SuccessCode = 0
/**
 * The code to exit an action with a "failure" state
 */
export const FailureCode = 1
/**
 * The code to exit an action with a "neutral" state
 */
export const NeutralCode = 78

export class Exit {
  /**
   * Stop the action with a "success" status
   */
  public success (message?: string) {
    if (message) console.log(message)
    process.exit(SuccessCode)
  }

  /**
   * Stop the action with a "neutral" status
   */
  public neutral (message?: string) {
    if (message) console.log(message)
    process.exit(NeutralCode)
  }

  /**
   * Stop the action with a "failed" status
   */
  public failure (message?: string) {
    if (message) console.error(message)
    process.exit(FailureCode)
  }
}
