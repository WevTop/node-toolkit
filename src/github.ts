import { graphql } from '@octokit/graphql'
import { Octokit } from '@octokit/rest'

export class GitHub extends Octokit {
  public graphql: typeof graphql

  constructor (token: string) {
    super({ auth: `token ${token}` })
    this.graphql = graphql.defaults({
      headers: { authorization: `token ${token}` }
    })
  }
}
