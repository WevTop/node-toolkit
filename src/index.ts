import Octokit from '@octokit/rest'
import fs from 'fs'
import path from 'path'
import execa from 'execa'
import yaml from 'js-yaml'
import Context from './context'

class Toolkit {
  public context: Context

  /**
   * Path to a clone of the repository
   */
  public workspace: string
  /**
   * GitHub API token
   */
  public token: string

  constructor () {
    // Print a console warning for missing environment variables
    this.warnForMissingEnvVars()

    this.context = new Context()
    this.workspace = process.env.GITHUB_WORKSPACE as string
    this.token = process.env.GITHUB_TOKEN as string
  }

  private warnForMissingEnvVars () {
    const requiredEnvVars = [
      'HOME',
      'GITHUB_WORKFLOW',
      'GITHUB_ACTION',
      'GITHUB_ACTOR',
      'GITHUB_REPOSITORY',
      'GITHUB_EVENT_NAME',
      'GITHUB_EVENT_PATH',
      'GITHUB_WORKSPACE',
      'GITHUB_SHA',
      'GITHUB_REF',
      'GITHUB_TOKEN'
    ]
    
    const requiredButMissing = requiredEnvVars.filter(key => process.env.hasOwnProperty(key))
    if (requiredButMissing.length > 0) {
      // This isn't being run inside of a GitHub Action environment!
      console.warn(`There are environment variables missing from this runtime, but would be present on GitHub.\n${requiredButMissing.map(key => `- ${key}`).join('\n')}`)
    }
  }

  /**
   * Returns an authenticated Octokit client.
   */
  createOctokit () {
    if (!this.token) {
      throw new Error('No `GITHUB_TOKEN` environment variable found, could not authenticate Octokit client.')
    }

    const octokit = new Octokit()

    octokit.authenticate({
      type: 'token',
      token: this.token
    })

    return octokit
  }

  /**
   * Gets a file in your project's workspace
   *
   * @param filename - Name of the file
   * @param encoding - Encoding (usually utf8)
   */
  getFile (filename: string, encoding = 'utf8') {
    const pathToFile = path.join(this.workspace, filename)
    if (!fs.existsSync(pathToFile)) throw new Error(`File ${filename} could not be found in your project's workspace.`)
    return fs.readFileSync(pathToFile, encoding)
  }

  /**
   * Get the package.json file in the project root
   * 
   * ```js
   * const pkg = toolkit.getPackageJSON()
   * ```
   */
  getPackageJSON () : object {
    const pathToPackage = path.join(this.workspace, 'package.json')
    if (!fs.existsSync(pathToPackage)) throw new Error('package.json could not be found in your project\'s root.')
    return require(pathToPackage)
  }

  /**
   * Get the configuration settings for this action in the project workspace.
   *
   * @param key - If this is a string like `.myfilerc` it will look for that file.
   * If it is a YAML file, it will return it as a JSON object. Otherwise, it will return the value of the property in
   * the `package.json` file of the project.
   * 
   * @example This method can be used in three different ways:
   *
   * ```js
   * // Get the .rc file
   * const cfg = toolkit.config('.myactionrc')
   *
   * // Get the YAML file
   * const cfg = toolkit.config('myaction.yml')
   *
   * // Get the property in package.json
   * const cfg = toolkit.config('myaction')
   * ```
   */
  config (key: string) : object {
    if (/\..+rc/.test(key)) {
      // It's a file like .npmrc or .eslintrc!
      const pathToRcFile = path.join(this.workspace, key)
      if (!fs.existsSync(pathToRcFile)) throw new Error(`File ${key} could not be found in your project's workspace.`)
      return require(pathToRcFile)
    } else if (key.endsWith('.yml') || key.endsWith('.yaml')) {
      // It's a YAML file! Gotta serialize it!
      return yaml.safeLoad(this.getFile(key))
    } else {
      // It's a regular object key in the package.json
      const pkg = this.getPackageJSON() as any
      return pkg[key]
    }
  }

  /**
   * Run a CLI command in the workspace
   *
   * @param command - Command to run
   * @param args - Arguments
   * @param cwd - Directory to run the command in
   * @param opts - 
   */
  async runInWorkspace (command: string, args: string[], cwd = this.workspace, opts: object) {
    return execa(command, args, { cwd, ...opts })
  }
}

export default Toolkit
module.exports = Toolkit