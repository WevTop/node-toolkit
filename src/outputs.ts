import * as core from '@actions/core'

export interface OutputType { [key: string]: string | undefined }

export function createOutputProxy <O extends OutputType = OutputType>() {
  return new Proxy<O>({} as O, {
    set (_, name: string, value: string): boolean {
      // When we attempt to get `outputs.___`, instead
      // we call `core.setOutput`.
      core.setOutput(name, value)
      return true;
    },
    getOwnPropertyDescriptor() {
      return {
        enumerable: false,
        configurable: true,
        writable: true
      }
    }
  })
}