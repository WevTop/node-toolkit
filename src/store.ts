import fs from 'fs'

export default class Store extends Map<string, any> {
  constructor (path: string) {
    if (fs.existsSync(path)) {
      const contents = fs.readFileSync(path, 'utf8')
      const json = JSON.parse(contents)
      super(json)
    } else {
      super()
    }

    process.on('exit', async () => {
      if (this.size === 0) return
      return this.write(path)
    })
  }

  public async write (path: string) {
    return new Promise((resolve, reject) => {
      const contents = JSON.stringify(Array.from(this))
      fs.writeFile(path, contents, (err) => {
        if (err) return reject(err)
        return resolve()
      })
    })
  }
}
