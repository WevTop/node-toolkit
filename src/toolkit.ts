import Octokit from '@octokit/rest'
import fs from 'fs'
import path from 'path'
import execa from 'execa'
import yaml from 'js-yaml'
import Context from './context'

export interface Workspace {
  path: string
}

export default class Toolkit {
  public context: Context
  public workspace: Workspace

  constructor () {
    this.context = new Context()
    this.workspace = {
      path: String(process.env.GITHUB_WORKSPACE)
    }
  }

  /**
   * Returns an authenticated Octokit client
   *
   * @returns {object}
   */
  createOctokit () {
    const octokit = new Octokit()

    octokit.authenticate({
      type: 'token',
      token: String(process.env.GITHUB_TOKEN)
    })

    return octokit
  }

  /**
   * Gets a file in your project's workspace
   *
   * @param {string} filename - Filename
   * @param {string} [encoding='utf8'] - Encoding
   *
   * @returns {string}
   */
  getFile (filename: string, encoding = 'utf8') : string {
    const pathToFile = path.join(this.workspace.path, filename)
    if (!fs.existsSync(pathToFile)) throw new Error(`File ${filename} could not be found in your project's workspace.`)
    return fs.readFileSync(pathToFile, encoding)
  }

  /**
   * Get the package.json file in the project root
   *
   * @returns {object}
   */
  getPackageJSON () {
    const pathToPackage = path.join(this.workspace.path, 'package.json')
    if (!fs.existsSync(pathToPackage)) throw new Error('package.json could not be found in your project\'s root.')
    return require(pathToPackage)
  }

  /**
   * Get the configuration settings for this action in the project workspace.
   *
   * @param {string} key - If this starts with a `.`, it will look for a file starting with `.`.
   * If it is a YAML file, it will return it as JSON. Otherwise, it will return the value of the property in
   * the `package.json` file of the project.
   *
   * @returns {object}
   */
  config (key: string) {
    if (key.startsWith('.') && key.endsWith('rc')) {
      // It's a file like .npmrc or .eslintrc!
      const pathToRcFile = path.join(this.workspace.path, key)
      if (!fs.existsSync(pathToRcFile)) throw new Error(`File ${key} could not be found in your project's workspace.`)
      return require(pathToRcFile)
    } else if (key.endsWith('.yml') || key.endsWith('.yaml')) {
      // It's a YAML file! Gotta serialize it!
      return yaml.safeLoad(this.getFile(key))
    } else {
      // It's a regular object key in the package.json
      const pkg = this.getPackageJSON()
      return pkg[key]
    }
  }

  /**
   * Run a CLI command in the workspace
   *
   * @param {string} command - Command to run
   * @param {string[]} args - Arguments
   * @param {string} [subdirectory] - Subdirectory to run the command in
   *
   * @returns {Promise<object>}
   */
  async runInWorkspace (command: string, args: string[], cwd = this.workspace.path, opts: object) {
    return execa(command, args, { cwd, ...opts })
  }
}
