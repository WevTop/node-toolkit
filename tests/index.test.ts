import nock from 'nock'
import path from 'path'
import { Signale } from 'signale'
import { Toolkit } from '../src'
import { NeutralCode } from '../src/exit'

describe('Toolkit', () => {
  let toolkit: Toolkit

  beforeEach(() => {
    // Mock the `process` event emitter to prevent memory
    // leaks on repeated calls in tests - used by Store.
    process.on = jest.fn()
    toolkit = new Toolkit({ logger: new Signale({ disabled: true }) })
  })

  describe('#github', () => {
    it('returns a GitHub client', () => {
      expect(toolkit.github).toBeInstanceOf(Object)
    })

    it('returns a GraphQL function on `.graphql`', async () => {
      expect(toolkit.github.graphql).toBeInstanceOf(Function)

      const scoped = nock('https://api.github.com')
        .post('/graphql').reply(200, { data: { errors: [] } })

      await toolkit.github.graphql('query { }')
      expect(scoped.isDone()).toBe(true)
    })
  })

  describe('#getFile', () => {
    it('gets the contents of a file', () => {
      const actual = toolkit.getFile('README.md')
      expect(actual).toMatchSnapshot()
    })

    it('gets the contents of a file with custom encoding', () => {
      const actual = toolkit.getFile('README.md', 'base64')
      expect(actual).toMatchSnapshot()
    })

    it('throws if the file could not be found', () => {
      const actual = () => toolkit.getFile('DONTREADME.md')
      expect(actual).toThrowErrorMatchingSnapshot()
    })
  })

  describe('#getPackageJSON', () => {
    it('returns the package.json file as a JSON object', () => {
      const actual = toolkit.getPackageJSON()
      expect(actual).toMatchSnapshot()
    })

    it('throws if the package.json file could not be found', () => {
      toolkit.workspace = path.join(__dirname, 'fixtures', 'workspaces', 'no-package-json')
      const actual = () => toolkit.getPackageJSON()
      expect(actual).toThrowErrorMatchingSnapshot()
    })
  })

  describe('#command', () => {
    it('calls the handler without any args', () => {
      const spy = jest.fn()
      toolkit.context.payload.comment = { body: '/action' }
      toolkit.command('action', spy)
      expect(spy).toHaveBeenCalled()
    })

    it('ignores commands not at the beginning of the line', () => {
      const spy = jest.fn()
      toolkit.context.payload.comment = { body: 'Hello /action' }
      toolkit.command('action', spy)
      expect(spy).not.toHaveBeenCalled()
    })

    it('only matches the exact command', () => {
      const spy = jest.fn()
      toolkit.context.payload.comment = { body: '/actionssssssssss' }
      toolkit.command('action', spy)
      expect(spy).not.toHaveBeenCalled()
    })

    it('calls the handler with a command at the beginning of a line that is not the first line', () => {
      const spy = jest.fn()
      toolkit.context.payload.comment = { body: 'Hello\n/action' }
      toolkit.command('action', spy)
      expect(spy).toHaveBeenCalled()
    })

    it('calls the handler with parsed args', () => {
      const spy = jest.fn()
      toolkit.context.payload.comment = { body: '/action testing another --file index.js' }
      toolkit.command('action', spy)
      expect(spy).toHaveBeenCalled()
      expect(spy).toHaveBeenCalledWith({
        _: ['testing', 'another'],
        file: 'index.js'
      }, expect.arrayContaining([
        '/action testing another --file index.js',
        'testing another --file index.js'
      ]))
    })

    it('does not call the handler if the body does not contain the command', () => {
      const spy = jest.fn()
      toolkit.context.payload.sender = { type: 'User' }
      toolkit.context.payload.comment = { body: 'Hello how are you' }
      toolkit.command('action', spy)
      expect(spy).not.toHaveBeenCalled()
    })

    it('does not call the handler if no body is found', () => {
      const spy = jest.fn()
      toolkit.command('action', spy)
      expect(spy).not.toHaveBeenCalled()
    })

    it('does not call the handler if the sender was a bot', () => {
      const spy = jest.fn()

      toolkit.context.payload.comment = { body: '/action' }
      toolkit.context.payload.sender = { type: 'Bot' }

      toolkit.command('action', spy)
      expect(spy).not.toHaveBeenCalled()
    })
  })

  describe('#config', () => {
    it('returns a property in the package.json', () => {
      const actual = toolkit.config('action')
      expect(actual).toEqual({ foo: true })
    })

    it('returns a parsed YAML file', () => {
      const actual = toolkit.config('action.yml')
      expect(actual).toEqual({ foo: true })
    })

    it('returns a .rc file as JSON', () => {
      const actual = toolkit.config('.actionrc')
      expect(actual).toEqual({ foo: true })
    })
  })

  describe('#runInWorkspace', () => {
    it('runs the command in the workspace', async () => {
      const result = await toolkit.runInWorkspace('echo', 'hello')
      expect(result).toMatchSnapshot()
      expect(result.stdout).toBe('hello')
    })

    it('runs the command in the workspace with some options', async () => {
      const result = await toolkit.runInWorkspace('throw', undefined, { reject: false })
      expect(result).toMatchSnapshot()
    })
  })

  describe('#wrapLogger', () => {
    it('wraps the provided logger and allows for a callable class', () => {
      const logger = new Signale() as jest.Mocked<Signale>
      logger.info = jest.fn()
      const twolkit = new Toolkit({ logger })

      twolkit.log('Hello!')
      twolkit.log.info('Hi!')

      expect(logger.info).toHaveBeenCalledTimes(2)
      expect(logger.info).toHaveBeenCalledWith('Hello!')
      expect(logger.info).toHaveBeenCalledWith('Hi!')
    })
  })
})

describe('Toolkit#constructor', () => {
  let logger: jest.Mocked<Signale>
  let exit: (code?: number) => never

  beforeEach(() => {
    logger = new Signale() as jest.Mocked<Signale>
    logger.error = jest.fn()
    logger.warn = jest.fn()

    exit = global.process.exit
    const p = global.process as any
    p.exit = jest.fn()
  })

  it('exits if the event is not allowed with an array of events', () => {
    // tslint:disable-next-line:no-unused-expression
    new Toolkit({ logger, event: ['pull_request'] })
    expect(process.exit).toHaveBeenCalledWith(NeutralCode)
    expect(logger.error.mock.calls).toMatchSnapshot()
  })

  it('does not exit if the event is one of the allowed with an array of events', () => {
    // tslint:disable-next-line:no-unused-expression
    new Toolkit({ logger, event: ['pull_request', 'issues'] })
    expect(process.exit).not.toHaveBeenCalled()
    expect(logger.error).not.toHaveBeenCalled()
  })

  it('exits if the event is not allowed with a single event', () => {
    // tslint:disable-next-line:no-unused-expression
    new Toolkit({ logger, event: 'pull_request' })
    expect(process.exit).toHaveBeenCalledWith(NeutralCode)
    expect(logger.error.mock.calls).toMatchSnapshot()
  })

  it('exits if the event is not allowed with an array of events with actions', () => {
    // tslint:disable-next-line:no-unused-expression
    new Toolkit({ logger, event: ['pull_request.opened'] })
    expect(process.exit).toHaveBeenCalledWith(NeutralCode)
    expect(logger.error.mock.calls).toMatchSnapshot()
  })

  it('exits if the event is not allowed with a single event with an action', () => {
    // tslint:disable-next-line:no-unused-expression
    new Toolkit({ logger, event: 'pull_request.opened' })
    expect(process.exit).toHaveBeenCalledWith(NeutralCode)
    expect(logger.error.mock.calls).toMatchSnapshot()
  })

  it('logs the expected string with missing env vars', () => {
    delete process.env.HOME
    new Toolkit({ logger }) // tslint:disable-line:no-unused-expression
    expect(logger.warn.mock.calls).toMatchSnapshot()
  })

  afterEach(() => {
    global.process.exit = exit
  })
})
