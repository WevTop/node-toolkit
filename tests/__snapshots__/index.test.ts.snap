// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolkit #createOctokit throws if there is no GITHUB_TOKEN environment variable 1`] = `"No \`GITHUB_TOKEN\` environment variable found, could not authenticate Octokit client."`;

exports[`Toolkit #getFile gets the contents of a file 1`] = `"I'm writing this on a plane ✈️"`;

exports[`Toolkit #getFile gets the contents of a file with custom encoding 1`] = `"SSdtIHdyaXRpbmcgdGhpcyBvbiBhIHBsYW5lIOKciO+4jw=="`;

exports[`Toolkit #getFile throws if the file could not be found 1`] = `"File DONTREADME.md could not be found in your project's workspace."`;

exports[`Toolkit #getPackageJSON returns the package.json file as a JSON object 1`] = `
Object {
  "action": Object {
    "foo": true,
  },
}
`;

exports[`Toolkit #getPackageJSON throws if the package.json file could not be found 1`] = `"package.json could not be found in your project's root."`;

exports[`Toolkit #runInWorkspace runs the command in the workspace 1`] = `
Object {
  "cmd": "echo hello",
  "code": 0,
  "failed": false,
  "killed": false,
  "signal": null,
  "stderr": "",
  "stdout": "hello",
  "timedOut": false,
}
`;

exports[`Toolkit #warnForMissingEnvVars logs the expected string 1`] = `
"There are environment variables missing from this runtime, but would be present on GitHub.
- HOME"
`;
